src/shell/shell_misc.cpp

```diff
@@ -462,17 +459,18 @@ bool DOS_Shell::Execute(std::string_view name, std::string_view args)
 	auto extension = fullname.substr(fullname.size() - extension_size);
 
 	if (iequals(extension, ".BAT")) {
-		/* delete old batch file if call is not active*/
-		/*keep the current echostate (as delete bf might change it )*/
-		const bool temp_echo = echo;
-		if (bf && !call) {
-			bf.reset();
+		if (!batchfiles.empty() && !call) {
+			batchfiles.pop();
+		}
+		try {
+			batchfiles.emplace(*this,
+			                   std::make_unique<FileReader>(fullname),
+			                   name,
+			                   args,
+			                   echo);
+		} catch (const std::exception&) {
+			WriteOut("Could not open %s", fullname.c_str());
```

Comment:
please avoid adding new exceptions, and when possible use non-throwing calls and check return codes (or use `std::optional<>` to return unavailable objects). 

For example, a file that can't be opened is a routine scenario and should be handled. It's typical and normal.  

Some C++ calls _do_ throw and we deliberately don't catch those when they're truly exceptional.  For example, if the host can't allocate memory, `new` will throw.  We let that exception go all the way to the top and take down the process, because there's no way to proceed from that.

Reply:
Yup; throwing and catching doesn't exist in DOSBox, preferring function return codes or member heath checks. (It's a good fit for DOS too, which was all return-value based).

Some of our most user-data-ladden object are the MIDI devices classes like FluidSynth and MT32: we need to load various ROMs, geting mt32emu running, get threading going, ensure connections to MPU-401, and so on.. 

If any of those fail, the MIDI device's `bool is_open` member never gets set positive, and the outside code can discard the object and inform the users.

Would an similar health-check work here too? `IsOpen()` or whatever makes sense for a new FileReader object. Then you can throw it away if didn't fully construct.

Reply:
You could instantiate it first as a bare object, and only std::move() it into the `batchfiles` container if it's healthy.

Reply:
Yeah, those are single device objects (the entire midi device, and aren't being blasted out by the hundreds) -- yeah; let's see how the wrapper function works! Maybe it'll be a better approach, and we can start to move towards it.